File: src/engine/adapters/sqs/catalog.ts
import { GetQueueUrlCommand, ListQueuesCommand, type SQSClient } from '@aws-sdk/client-sqs';
import { encodePath, type TreeNode } from '@shared/domain/tree';
import { mapError } from './errors';

const PAGE_LIMIT = 1000; // ListQueues's own max MaxResults per call

export interface QueueListing {
  nodes: TreeNode[];
  /** Name -> URL, surfaced so index.ts's queueUrls cache can populate itself for free (D14). */
  urlByName: Map<string, string>;
}

// §5.1's sqs row: tree is a flat queue list, no deeper level. `name` is the URL's last path
// segment (the queue name); listQueues already has every queue's full URL in hand while paging,
// so it hands the name->URL map back instead of discarding it — index.ts caches it, avoiding a
// GetQueueUrl round trip on every read()/count() (D14, fixes F14/F22).
export async function listQueues(client: SQSClient): Promise<QueueListing> {
  const nodes: TreeNode[] = [];
  const urlByName = new Map<string, string>();
  let nextToken: string | undefined;
  do {
    let result: { QueueUrls?: string[]; NextToken?: string };
    try {
      result = await client.send(
        new ListQueuesCommand({ MaxResults: PAGE_LIMIT, NextToken: nextToken }),
      );
    } catch (err) {
      throw mapError(err);
    }
    for (const url of result.QueueUrls ?? []) {
      const name = url.slice(url.lastIndexOf('/') + 1);
      urlByName.set(name, url);
      nodes.push({
        kind: 'queue',
        name,
        path: encodePath([{ kind: 'queue', name }]),
        hasChildren: false,
      });
    }
    nextToken = result.NextToken;
  } while (nextToken);

  nodes.sort((a, b) => a.name.localeCompare(b.name));
  return { nodes, urlByName };
}

/** Resolves a queue's leaf `name` back to its full URL — GetQueueUrl is a single cheap call. */
export async function resolveQueueUrl(client: SQSClient, name: string): Promise<string> {
  try {
    const result = await client.send(new GetQueueUrlCommand({ QueueName: name }));
    if (!result.QueueUrl) throw new Error(`queue not found: ${name}`);
    return result.QueueUrl;
  } catch (err) {
    throw mapError(err);
  }
}

File: shell/internal/adapters/sqs/catalog.go
package sqs

import (
	"context"
	"fmt"
	"sort"
	"strings"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/sqs"

	"github.com/kirathecat/kira-studio/shell/internal/storage/model"
)

const pageLimit = 1000 // ListQueues's own max MaxResults per call

// queueListing is catalog.ts's QueueListing.
type queueListing struct {
	nodes     []model.TreeNode
	urlByName map[string]string
}

// listQueues is catalog.ts's listQueues. The tree is a flat queue list, no deeper level; name is
// the URL's last path segment. ListQueues already has every queue's full URL in hand while paging,
// so it hands the name->URL map back instead of discarding it — adapter.go caches it, avoiding a
// GetQueueUrl round trip on every read()/count() call (P58d D9).
func listQueues(ctx context.Context, client *sqs.Client) (queueListing, error) {
	nodes := []model.TreeNode{}
	urlByName := map[string]string{}
	var nextToken *string
	for {
		result, err := client.ListQueues(ctx, &sqs.ListQueuesInput{MaxResults: aws.Int32(pageLimit), NextToken: nextToken})
		if err != nil {
			return queueListing{}, mapError(err)
		}
		for _, url := range result.QueueUrls {
			name := url
			if i := strings.LastIndexByte(url, '/'); i >= 0 {
				name = url[i+1:]
			}
			urlByName[name] = url
			nodes = append(nodes, model.TreeNode{
				Kind: "queue", Name: name, Path: model.EncodePath([]model.PathSegment{{Kind: "queue", Name: name}}),
				HasChildren: false,
			})
		}
		nextToken = result.NextToken
		if nextToken == nil {
			break
		}
	}
	sort.Slice(nodes, func(i, j int) bool { return nodes[i].Name < nodes[j].Name })
	return queueListing{nodes: nodes, urlByName: urlByName}, nil
}

// resolveQueueURL is catalog.ts's resolveQueueUrl — a queue's leaf name resolved back to its full
// URL via a single cheap GetQueueUrl call.
func resolveQueueURL(ctx context.Context, client *sqs.Client, name string) (string, error) {
	result, err := client.GetQueueUrl(ctx, &sqs.GetQueueUrlInput{QueueName: aws.String(name)})
	if err != nil {
		return "", mapError(err)
	}
	if result.QueueUrl == nil {
		return "", mapError(fmt.Errorf("queue not found: %s", name))
	}
	return *result.QueueUrl, nil
}
